spinoso-time/src/time/tzrs/offset.rs

```diff
@@ -0,0 +1,273 @@
+///
+/// Note: this matches MRI Ruby implmentation. Where `TZ="" ruby -e "puts Time::now"` will return a
+/// new _time_ with 0 offset from UTC, but still still report as a non utc time.
+///
+/// [`local_tz`]: https://docs.rs/tzdb/latest/tzdb/fn.local_tz.html
+/// [`time_zone_designation`]: https://docs.rs/tz-rs/0.6.9/tz/timezone/struct.LocalTimeType.html#method.time_zone_designation
+#[inline]
+#[must_use]
+fn local_time_zone() -> TimeZoneRef<'static> {
+    match local_tz() {
+        Some(tz) => tz,
+        None => GMT,
+    }
+}
+
+/// Generates a [+/-]HHMM timezone format from a given number of seconds
+/// Note: the actual seconds element is effectively ignored here
+#[inline]
+#[must_use]
+fn offset_hhmm_from_seconds(seconds: i32) -> String {
+    let flag = if seconds < 0 { '-' } else { '+' };
+    let minutes = seconds.abs() / 60;
+
+    let offset_hours = minutes / 60;
+    let offset_minutes = minutes - (offset_hours * 60);
+
+    format!("{}{:0>2}{:0>2}", flag, offset_hours, offset_minutes)
+}
+
+/// Represents the number of seconds offset from UTC
+#[allow(variant_size_differences)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Offset {
+    /// UTC offset, zero offset, Zulu time
+    Utc,
+    /// Fixed offset from UTC
+    ///
+    /// Note: A fixed offset of 0 is different from UTC time
+    Fixed([LocalTimeType; 1]),
+    /// A time zone based offset
+    Tz(TimeZoneRef<'static>),
+}
+
+impl<'a> Offset {
+    /// Generate a UTC based offset
+    #[inline]
+    #[must_use]
+    pub fn utc() -> Self {
+        Self::Utc
+    }
+
+    /// Generate an offset based on the detected local time zone of the system
+    ///
+    /// Detection is done by [`tzdb::local_tz`], and if it fails will return a GMT timezone
+    ///
+    /// [`tzdb::local_tz`]: https://docs.rs/tzdb/latest/tzdb/fn.local_tz.html
+    #[inline]
+    #[must_use]
+    pub fn local() -> Self {
+        Self::Tz(local_time_zone())
+    }
+
+    /// Generate an offset with a number of seconds from UTC.
+    #[inline]
+    #[must_use]
+    pub fn fixed(offset: i32) -> Self {
+        let offset_name = offset_hhmm_from_seconds(offset);
+        let local_time_type =
+            LocalTimeType::new(offset, false, Some(offset_name.as_bytes())).expect("Couldn't create fixed offset");
+        Self::Fixed([local_time_type])
+    }
+
+    /// Generate an offset based on a provided [`tz::timezone::TimeZoneRef`]
+    ///
+    /// This can be combined with [`tzdb`] to generate offsets based on predefined iana time zones
+    ///
+    /// ```
+    /// use spinoso_time::tzrs::Offset;
+    /// use tzdb::time_zone::pacific::AUCKLAND;
+    /// let offset = Offset::tz(AUCKLAND);
+    /// ```
+    ///
+    /// [`tz:timezone::TimeZoneRef`]: https://docs.rs/tz-rs/0.6.9/tz/timezone/struct.TimeZoneRef.html
+    /// [`tzdb`]: https://docs.rs/tzdb/latest/tzdb/index.html
+    #[inline]
+    #[must_use]
+    pub fn tz(tz: TimeZoneRef<'static>) -> Self {
```

Comment:
if `TimeZoneRef` is part of `spinoso-time`'s public API, let's make sure we export the type from `spinoso_time::tzrs` with something like:

```rust
#[doc(inline)]
pub use tz::timezone::TimeZoneRef;
```

Reply:
I was thinking about this one. How do you feel about only exposing a `&[u8]` interface instead?

I should mention that MRI doesn't come with any timezone data (I believe it only knows about the local timezone via sys calls). Adding a timezone db functionality + Timezone by name is funnily enough more than what MRI supports. The reason for a &[u8] interface is to support a "lookup by name" approach.

Note: MRI also supports a ruby object with the relevant timezone timezone fields (too hard to support in spinoso-time, could be ruby functions etc)

Reply:
Exposing a `&[u8]` interface works for me. If that's the case we'll want to rework the `Offset` enum to hide its guts:

```rust
/// Represents the number of seconds offset from UTC
pub struct Offset(OffsetType);

/// Represents the number of seconds offset from UTC
#[allow(variant_size_differences)]
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
enum OffsetType {
    /// UTC offset, zero offset, Zulu time
    Utc,
    /// Fixed offset from UTC
    ///
    /// Note: A fixed offset of 0 is different from UTC time
    Fixed([LocalTimeType; 1]),
    /// A time zone based offset
    Tz(TimeZoneRef<'static>),
}
```

This is kinda the same pattern:

https://github.com/artichoke/artichoke/blob/a9a3bb61351f6c81286e3122b0ab53f65907f6e1/spinoso-string/src/enc/inspect.rs#L7

https://github.com/artichoke/artichoke/blob/a9a3bb61351f6c81286e3122b0ab53f65907f6e1/spinoso-string/src/enc/inspect.rs#L30-L57

Reply:
![it's always enums](https://i.imgflip.com/6kwpyf.jpg)

but in all seriousness, I think this makes a lot of sense. Something bugged me about exposing the internals of tz-rs to the outside world, and this would make things a lot more encapsulated.